DOC-4728

jamiltz opened the pull request:
https://issues.couchbase.com/browse/DOC-4728

- adds a search page for tutorials (`/tutorials.html`)
- expands on the existing `.tiles` CSS class
- adds CSS specific to the tutorials page (`tutorials.css`)
- adds `mark.js` as a dependency for the search/highlighting feature

jamiltz commented:
@mojavelinux I think this PR is ready for review (I've written a summary of changes on the first PR comment).
Would be great to your review on it.
Preview src page: https://deploy-preview-49--cb-docs-ui.netlify.com/tutorials.html

mojavelinux commented:
@jamiltz I think this is a great start. I left some comments in the review. My main concern, however, is that the style isn't quite consistent with the brand guidelines. It mostly has to do with the spacing / padding, though the font sizes seem a bit off too. I would strongly recommend that you have the web design team look this over and make adjustments to the styles so that it aligns more with the .com site, for example https://www.couchbase.com/partners/find-a-partner. Ideally, the experience should be seamless. And we should expect that they are involved, because this is their area of expertise.

jamiltz commented on src/partials/footer-scripts.hbs:
```diff
@@ -1,3 +1,4 @@
+<script src="https://cdn.jsdelivr.net/mark.js/8.6.0/mark.min.js"></script>
```
https://github.com/couchbase/docs-ui/pull/49/commits/906f65e52f447c00bb424e83d263f58a64015ae5

mojavelinux replied:
Excellent. Though you're missing the bundle step in the build task. I'll notate the line.

jamiltz commented on src/css/doc.css:
```diff
@@ -688,6 +688,142 @@
   font-weight: var(--weight-bold);
 }
 
+/** CSS rules for tiles on the tutorial list page **/
```
OK. I've moved the changes in `doc.css` over to `home.css` https://github.com/couchbase/docs-ui/pull/49/commits/4408b32a29ef2db8ada5a15a0060e933d9ec0286

jamiltz commented on src/partials/body-tutorials.hbs:
```diff
@@ -0,0 +1,38 @@
+                        autofocus
+                        class="query"
+                        id="search-tutorials"
+                        placeholder="Search Couchbase Tutorials"
```
Good point, I've updated the placeholder https://github.com/couchbase/docs-ui/pull/49/commits/2ea713b4d41ff0983700efdbd4f2ab78b56ccdc6

mojavelinux replied:
There's still an extra "s". It should be "Find a Couchbase Tutorial" instead of "Find a Couchbase Tutorials".

jamiltz commented on src/partials/body-tutorials.hbs:
```diff
@@ -0,0 +1,38 @@
+            </div>
+          </div>
+        </div>
+      </nav>
```
Fixed https://github.com/couchbase/docs-ui/pull/49/commits/4595082a444834e7df19b8d1a14586c855726532

jamiltz commented on src/layouts/tutorials.hbs:
```diff
@@ -0,0 +1,16 @@
+  {{> head-first}}
+  <title>Couchbase Tutorials</title>
+  {{#if site.url}}
+    <link rel="canonical" href="{{{site.url}}}">
```
Fixed https://github.com/couchbase/docs-ui/commit/4595082a444834e7df19b8d1a14586c855726532

jamiltz commented on src/css/site.css:
```diff
@@ -17,3 +17,4 @@
 @import "optanon.css";
 @import "highlight.css";
 @import "feedback.css";
+@import "tutorials.css";
```
Yes good point, done in https://github.com/couchbase/docs-ui/pull/49/commits/b748a7527cb72fdd1a3fa067b1837b8bb0701a73

jamiltz commented on preview-site-src/tutorials.adoc:
```diff
@@ -0,0 +1,182 @@
+
+[.developer]
+== Developer
+[.title]
```
Yes I usually do as well. For some reason I thought `docs-ui` asciidoc files would be different but there's not reason for that to be. Fixed in https://github.com/couchbase/docs-ui/pull/49/commits/27afb5fdfbed0341f31465ecf94e92caf2c71e6f

jamiltz commented on preview-site-src/tutorials.adoc:
```diff
@@ -0,0 +1,182 @@
+[.architect]
+== Architect
+[.title]
+xref:tutorials:mobile-travel-sample:swift/design/data-modeling.adoc[Deploy Couchbase in Kubernetes]
```
Yes! I had not thought about making the xref a discrete heading. I've indented other headers accordingly. https://github.com/couchbase/docs-ui/pull/49/commits/20d2f31260143c1175a1e8926ab09bd1fb7c1f1a

jamiltz commented:
@mojavelinux Thanks for the thorough review! I've addressed the review comments, received the web team's recommendations and applied them.

I'd appreciate a review of the fix to bundle `mark.js` in the site output (commit https://github.com/couchbase/docs-ui/commit/906f65e52f447c00bb424e83d263f58a64015ae5#diff-e1d72f8220e801ddc8e7445a441626ef). Is that how it should be done? Thanks!

jamiltz commented:
@mojavelinux is there a way in Asciidoc to create a new parent div in the output? I'm currently using `===== {empty}` which outputs the following (with an empty `h5` element):

```
<div class="sect5">
  <h5></h5>
</div>
```

mojavelinux commented:
@jamiltz I'm looking over the updates now. From your comments, I anticipate it's already in great shape!

> @mojavelinux is there a way in Asciidoc to create a new parent div in the output? I'm currently using `===== {empty}` which outputs the following (with an empty `h5` element):

That's certainly one way. Another is an open block, though that can conflict with nested / included content. Yet another solution is to use an example block with a delimiter length of more than 4 (such as 6) so it doesn't conflict with nested / included content.

Ultimately, this is the solution we want: https://github.com/asciidoctor/asciidoctor/issues/1121. But there are reasons why that has to remain on hold.

mojavelinux commented on src/layouts/tutorials.hbs:
```diff
@@ -0,0 +1,16 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  {{> head-first}}
```
Handlebars includes should always be aligned with the left margin (to avoid the content it includes from being indented).

```
<head>
{{> head-first}}
  <title>Couchbase Tutorials</title>
```

jamiltz replied:
OK I'll try to remember from now on. Fixed in https://github.com/couchbase/docs-ui/pull/49/commits/4b2305f591514f502403aa891374596a1310754e

mojavelinux commented on src/partials/body-tutorials.hbs:
```diff
@@ -0,0 +1,38 @@
+  <div class="home" data-ceiling="topbar">
+    <div class="tutorials">
+      <article class="doc">
+        {{{ page.contents }}}
```
Don't include space inside the Handlebars. Also, don't indent included content. Otherwise, Handlebars likes to add that indentation to every line (unless a special option is passed to Handlebars, which may or may not be set).

```html
      <article class="doc">
{{{page.contents}}}
      </article>
```

jamiltz replied:
Fixed in https://github.com/couchbase/docs-ui/pull/49/commits/0d0f4698e8eccf91c6487a3dbd8ffe5285eb550d

mojavelinux commented on src/partials/footer-scripts.hbs:
```diff
@@ -11,3 +11,4 @@
 {{#if (eq env.FEEDBACK_BUTTON 'true')}}
 <script async id="feedback-script" src="{{uiRootPath}}/js/vendor/feedback.js?v=1" data-collector-id="de2e9313"></script>
 {{/if}}
+<script async src="{{uiRootPath}}/js/vendor/mark.js"></script>
```
You also need to add mark.js to tasks/build.js

```js
    vfs
      .src('js/vendor/{docsearch,highlight,mark}.js', Object.assign({ read: false }, opts))
```

jamiltz replied:
Fixed in https://github.com/couchbase/docs-ui/pull/49/commits/98be6f49fcbb4da174ea5faa761fc70d4ef742dd

mojavelinux commented on src/partials/body-tutorials.hbs:
```diff
@@ -0,0 +1,38 @@
+        <div class="searchbar-menu">
+          <div class="searchbar-start">
+            <form class="searchbar-item search">
+              <input
```
One idea that I think would really improve the UX is to automatically scroll the search box to the top of the visible area when the user starts to type in the search field. This will leave room for the results below (without having to fix the scrolling of the page).

jamiltz replied:
Good idea! I've filed a ticket (as an improvement) to keep that feedback in mind https://issues.couchbase.com/browse/DOC-4815. Do you have a recommendation for how to do that on our site? (should I use vanilla JavaScript?)

mojavelinux replied:
I would definitely use vanilla JavaScript (as it's pretty good these days). I would catch the first keyup event after focus, compute the offset top of the search box, then use window.scrollTo() to get there.

jamiltz replied:
Great, I've added the suggestion to the JIRA ticket. Thank you.

mojavelinux commented on src/css/home.css:
```diff
@@ -244,3 +244,243 @@
+  margin: 0 0 20px 0;
+  /* override property defined in .body.tiles .sect1 */
+  padding-right: 0;
+  margin-right: 20px;
```
I strongly recommend against using px units. Not only do rems (or ems) lend themselves better to scaling, they also encourage the use of consistent values. It's too tempting with px to just use any number, then you end up with sizes and spaces that don't have stepwise gradation.

jamiltz replied:
Ok good to know! Converted `px`s to `rems` in https://github.com/couchbase/docs-ui/pull/49/commits/e7d275c564fe276f20965be40f58dd510bbda719

mojavelinux commented on src/css/home.css:
```diff
@@ -244,3 +244,243 @@
+}
+
+.tutorials.body.tiles .sect1 {
+  border: 1px solid lightgrey;
```
I'm still not convinced about the design of the tiles. The lightgrey border especially sets of alarm bells for me because that's not coming from the palette of colors defined in vars.css (and it's also built-in color, which is almost guaranteed to be off brand).

I think the tiles should build on the design from one of these:

* persona tiles on https://docs.couchbase.com/home/index.html
* Featured Partners on https://www.couchbase.com/partners/find-a-partner
* Features on https://www.couchbase.com/products/developer-sdk
* Solve your toughest NoSQL adoption and production challenges on https://www.couchbase.com/products/n1ql

The persona tiles would be my recommendation.

I do understand that you have more content to display than in those tiles. So there's certainly some augmentation that needs to happen. But when you design the AsciiDoc, try to make sure to maintain the integrity of the AsciiDoc without making it cater too much to the HTML output. You can then use CSS to move content around.

For example, I think "Components" and "Languages" should be a description list, not section titles. That would then eliminate the need for the blank parent section.

jamiltz replied:
> I'm still not convinced about the design of the tiles. The lightgrey border especially sets of alarm bells for me because that's not coming from the palette of colors defined in vars.css (and it's also built-in color, which is almost guaranteed to be off brand).

Good point, I've changed it to the official color https://github.com/couchbase/docs-ui/pull/49/commits/545c017b521b87f0232afa1b5ca3bcae04d8a14b

> I think the tiles should build on the design from one of these:
>
> * persona tiles on https://docs.couchbase.com/home/index.html
> * Featured Partners on https://www.couchbase.com/partners/find-a-partner
> * Features on https://www.couchbase.com/products/developer-sdk
> * Solve your toughest NoSQL adoption and production challenges on https://www.couchbase.com/products/n1ql
>
> The persona tiles would be my recommendation.
>
> I do understand that you have more content to display than in those tiles. So there's certainly some augmentation that needs to happen. But when you design the AsciiDoc, try to make sure to maintain the integrity of the AsciiDoc without making it cater too much to the HTML output. You can then use CSS to move content around.

Yes, I think we should consider aligning with the persona tiles. Will mention it to the docs team. I've filed a ticket to track further https://issues.couchbase.com/browse/DOC-4817

> For example, I think "Components" and "Languages" should be a description list, not section titles. That would then eliminate the need for the blank parent section.

Filed a ticket to re-work it https://issues.couchbase.com/browse/DOC-4816

mojavelinux commented:
I left a review comment about how you can remove at least one of the blank sections. Instead of a blank section around the Components and Languages, you can use a description list.

jamiltz commented:
> I left a review comment about how you can remove at least one of the blank sections. Instead of a blank section around the Components and Languages, you can use a description list (though you may still run into a problem that description list items themselves aren't wrapped in a containing element, so they can be hard to layout).

OK. I've added that feedback to https://issues.couchbase.com/browse/DOC-4816 (I'm not sure how to layout Components and Languages if there is no containing element)

jamiltz commented:
Thanks for the re-review. I've addressed the change requests and filed JIRA tickets where I will need more time. I'll merge the PR 🎉

amarantha-k commented:
looks good. thank you!
